### PR TITLE
Allow for pgAdmin4 Storage to be Configurable

### LINKED ISSUE #1795: Separate parameter for pgadmin PVC size
When creating a pgadmin instance for a cluster, the PVC is taken from the cluster PVC size, which causes a large waste of space if you deploy the pgadmin module.

**Describe the solution you'd like**
Please add a separate parameter in the config to allow a separate fixed side for the pgadmin PVC.
Please also allow for an override on the pgo client.

**Please tell us about your environment:**

* Operating System: Centos 7
* Where is this running: Bare Metal Local
* Storage being used: Rook-Ceph
* Container Image Tag: V4.4.0
* PostgreSQL Version: default (v12)
* Platform: K8s
* Platform Version: 1.18

### PR DESCRIPTION
**Pull request by tjmoore4**
**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change, as applicable?
 - [x] Have you tested your changes on all related environments with successful results, as applicable?

**Type of Changes:**

 
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**
Currently, the pgAdmin4 PVC relies on the same storage
system as a PostgreSQL primary, which, for some use
cases, will not fit the required use case.

**What is the new behavior (if this is a feature change)?**
This update allows for the pgAdmin4 storage configuration
to be customized. It adds a '--storage-config' and
'--pvc-size' to the 'pgo create pgadmin' command, as well as
a 'pgadmin_storage' option to the various install configuration
files. The values, if provided, are validated. If no values
are provided, the default primary storage configuration is used
as previously.

**Other information**:
Issue: [ch8067]
closes #1795

### CODE REVIEW

File: installers/helm/values.yaml [hunk 1]
```diff
@@ -92,6 +92,7 @@ backrest_storage: "default"
 backup_storage: "default"
 primary_storage: "default"
 replica_storage: "default"
+pgadmin_storage: ""
```
Comment (jkatz) [inline installers/helm/values.yaml:95]:
```suggestion
pgadmin_storage: "default"
```

File: internal/apiserver/pgadminservice/pgadminimpl.go [hunk 2]
```diff
@@ -61,14 +61,62 @@ func CreatePgAdmin(request *msgs.CreatePgAdminRequest, ns, pgouser string) msgs.
...
+		}
+
+		// validate the pgAdmin storage config value in the main configuration if the request did not provide one
+		// if it is set but not valid, return now
+		if request.StorageConfig == "" && apiserver.Pgo.PGAdminStorage != "" &&
+			!apiserver.IsValidStorageName(apiserver.Pgo.PGAdminStorage) {
+			resp.Status.Code = msgs.Error
+			resp.Status.Msg = fmt.Sprintf("%q configured storage config invalid", apiserver.Pgo.PGAdminStorage)
+			return resp
+		}
```
Comment (jkatz) [inline internal/apiserver/pgadminservice/pgadminimpl.go:85]: Shouldn't we validate this in `pgoconfig` when loading the configuration?

File: internal/apiserver/pgadminservice/pgadminimpl.go [hunk 3]
```diff
@@ -61,14 +61,62 @@ func CreatePgAdmin(request *msgs.CreatePgAdminRequest, ns, pgouser string) msgs.
...
+
+		// if the PVCSize is set to a customized value, ensure that it is recognizable by Kubernetes
+		if err := apiserver.ValidateQuantity(request.PVCSize); request.PVCSize != "" && err != nil {
+			resp.Status.Code = msgs.Error
+			resp.Status.Msg = fmt.Sprintf(apiserver.ErrMessagePVCSize, request.PVCSize, err.Error())
+			return resp
+		}
```
Comment (jkatz) [inline internal/apiserver/pgadminservice/pgadminimpl.go:92]: I would move this into the `if request.PVCSize != ""` block to keep it all together, and to ensure that we've set the StorageConfig at that point.

File: internal/config/pgoconfig.go [hunk 4]
```diff
@@ -323,6 +324,10 @@ func (c *PgoConfig) Validate() error {
 		if _, ok := c.Storage[c.ReplicaStorage]; !ok {
 			return storageNotDefined("ReplicaStorage", c.ReplicaStorage)
 		}
+		if _, ok := c.Storage[c.PGAdminStorage]; !ok {
+			log.Warning("PGAdminStorage setting not set, will use PrimaryStorage setting")
+			c.Storage[c.BackrestStorage] = c.Storage[c.PrimaryStorage]
```
Comment (jkatz) [inline internal/config/pgoconfig.go:329]:
```suggestion
			c.Storage[c.PGAdminStorage] = c.Storage[c.PrimaryStorage]
```

**Review by jkatz — COMMENTED** [review]
Looks good, I think this is just about there.

File: internal/apiserver/pgadminservice/pgadminimpl.go (hunk 2 above)
Comment (tjmoore4) [reply internal/apiserver/pgadminservice/pgadminimpl.go:85]: Yes, and there is a check there. This one is redundant, so I'll remove it.

File: installers/helm/values.yaml (hunk 1 above)
Comment (tjmoore4) [reply installers/helm/values.yaml:95]: updated.

File: internal/apiserver/pgadminservice/pgadminimpl.go (hunk 3 above)
Comment (tjmoore4) [reply internal/apiserver/pgadminservice/pgadminimpl.go:92]: makes sense to me, updated.

File: internal/config/pgoconfig.go (hunk 4 above)
Comment (tjmoore4) [reply internal/config/pgoconfig.go:329]: thanks for the catch, updated.

File: internal/apiserver/pgadminservice/pgadminimpl.go [hunk 5]
```diff
@@ -61,14 +61,60 @@ func CreatePgAdmin(request *msgs.CreatePgAdminRequest, ns, pgouser string) msgs.
...
+
+		// validate the pgAdmin storage config value in the main configuration if the request did not provide one
+		// if it is set but not valid, return now
+		if request.StorageConfig == "" && !apiserver.IsValidStorageName(apiserver.Pgo.PGAdminStorage) {
+			resp.Status.Code = msgs.Error
+			resp.Status.Msg = fmt.Sprintf("%q configured storage config invalid", apiserver.Pgo.PGAdminStorage)
+			return resp
+		}
```
Comment (jkatz) [inline internal/apiserver/pgadminservice/pgadminimpl.go:84]: I don't think we need this check if we've already validated the `apiserver.Pgo.PGAdminStorage` when we loaded the configuration.

File: internal/apiserver/pgadminservice/pgadminimpl.go [hunk 6]
```diff
@@ -61,14 +61,60 @@ func CreatePgAdmin(request *msgs.CreatePgAdminRequest, ns, pgouser string) msgs.
...
+
+		// Extract parameters for the optional pgAdmin storage. server configuration and
+		// request parameters are all optional.
+		// Now, set the main configured value, if provided
+		if apiserver.Pgo.PGAdminStorage != "" {
+			cluster.Spec.PGAdminStorage, _ = apiserver.Pgo.GetStorageSpec(apiserver.Pgo.PGAdminStorage)
+		}
```
Comment (jkatz) [inline internal/apiserver/pgadminservice/pgadminimpl.go:91]: We can just set this by default. If the user picked a one off, it's handled in the statement below this.

```suggestion
		cluster.Spec.PGAdminStorage, _ = apiserver.Pgo.GetStorageSpec(apiserver.Pgo.PGAdminStorage)
```

**Review by jkatz — COMMENTED** [review]
Couple of more items, and then outside of some documentation this is good to go.

File: internal/apiserver/pgadminservice/pgadminimpl.go (hunk 5 above)
Comment (tjmoore4) [reply internal/apiserver/pgadminservice/pgadminimpl.go:84]: yes, it is redundant at this point. It's been removed.

File: internal/apiserver/pgadminservice/pgadminimpl.go (hunk 6 above)
Comment (tjmoore4) [reply internal/apiserver/pgadminservice/pgadminimpl.go:91]: sure thing, updated.

File: internal/apiserver/pgadminservice/pgadminimpl.go [hunk 7]
```diff
@@ -61,14 +61,58 @@ func CreatePgAdmin(request *msgs.CreatePgAdminRequest, ns, pgouser string) msgs.
...
+			return resp
+		}
+
+		// validate the pgAdmin storage config value in the main configuration if the request did not provide one
+		// if it is set but not valid, return now
+		if request.StorageConfig == "" {
+			resp.Status.Code = msgs.Error
+			resp.Status.Msg = fmt.Sprintf("%q configured storage config invalid", apiserver.Pgo.PGAdminStorage)
+			return resp
+		}
```
Comment (jkatz) [inline internal/apiserver/pgadminservice/pgadminimpl.go:84]: Again, we don't need this block at all. We can just remove it.

Comment (tjmoore4) [reply internal/apiserver/pgadminservice/pgadminimpl.go]: yes, understood. Updated, thanks.

**Review by jkatz — APPROVED** [review]

**Pull request merged**